File: src/backend/estoque/src/model/estoque_reajuste_cabecalho_model.py
from src import db
from src.model.estoque_reajuste_detalhe_model import EstoqueReajusteDetalheModel
from src.model.view_pessoa_colaborador_model import ViewPessoaColaboradorModel


class EstoqueReajusteCabecalhoModel(db.Model):
    __tablename__ = 'estoque_reajuste_cabecalho'

    id = db.Column(db.Integer, primary_key=True)
    data_reajuste = db.Column(db.DateTime)
    taxa = db.Column(db.Float)
    tipo_reajuste = db.Column(db.String(1))
    justificativa = db.Column(db.String(100))
    id_colaborador = db.Column(db.Integer, db.ForeignKey('view_pessoa_colaborador.id'))

    estoque_reajuste_detalhe_model_list = db.relationship('EstoqueReajusteDetalheModel', lazy='dynamic')
    view_pessoa_colaborador_model = db.relationship('ViewPessoaColaboradorModel', foreign_keys=[id_colaborador])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_colaborador = data.get('idColaborador')
        self.data_reajuste = data.get('dataReajuste')
        self.taxa = data.get('taxa')
        self.tipo_reajuste = data.get('tipoReajuste')
        self.justificativa = data.get('justificativa')

    def serialize(self):
        return {
            'id': self.id,
            'idColaborador': self.id_colaborador,
            'dataReajuste': self.data_reajuste.isoformat(),
            'taxa': self.taxa,
            'tipoReajuste': self.tipo_reajuste,
            'justificativa': self.justificativa,
            'estoqueReajusteDetalheModelList': [estoque_reajuste_detalhe_model.serialize() for estoque_reajuste_detalhe_model in self.estoque_reajuste_detalhe_model_list],
            'viewPessoaColaboradorModel': self.view_pessoa_colaborador_model.serialize() if self.view_pessoa_colaborador_model else None,
        }

File: frontend/estoque/lib/app/data/model/estoque_reajuste_cabecalho_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:estoque/app/infra/infra_imports.dart';
import 'package:estoque/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';
import 'package:estoque/app/data/domain/domain_imports.dart';

class EstoqueReajusteCabecalhoModel {
	int? id;
	int? idColaborador;
	DateTime? dataReajuste;
	double? taxa;
	String? tipoReajuste;
	String? justificativa;
	List<EstoqueReajusteDetalheModel>? estoqueReajusteDetalheModelList;
	ViewPessoaColaboradorModel? viewPessoaColaboradorModel;

	EstoqueReajusteCabecalhoModel({
		this.id,
		this.idColaborador,
		this.dataReajuste,
		this.taxa,
		this.tipoReajuste,
		this.justificativa,
		this.estoqueReajusteDetalheModelList,
		this.viewPessoaColaboradorModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'data_reajuste',
		'taxa',
		'tipo_reajuste',
		'justificativa',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Data Reajuste',
		'Taxa',
		'Tipo Reajuste',
		'Justificativa',
	];

	EstoqueReajusteCabecalhoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idColaborador = jsonData['idColaborador'];
		dataReajuste = jsonData['dataReajuste'] != null ? DateTime.tryParse(jsonData['dataReajuste']) : null;
		taxa = jsonData['taxa']?.toDouble();
		tipoReajuste = EstoqueReajusteCabecalhoDomain.getTipoReajuste(jsonData['tipoReajuste']);
		justificativa = jsonData['justificativa'];
		estoqueReajusteDetalheModelList = (jsonData['estoqueReajusteDetalheModelList'] as Iterable?)?.map((m) => EstoqueReajusteDetalheModel.fromJson(m)).toList() ?? [];
		viewPessoaColaboradorModel = jsonData['viewPessoaColaboradorModel'] == null ? ViewPessoaColaboradorModel() : ViewPessoaColaboradorModel.fromJson(jsonData['viewPessoaColaboradorModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idColaborador'] = idColaborador != 0 ? idColaborador : null;
		jsonData['dataReajuste'] = dataReajuste != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataReajuste!) : null;
		jsonData['taxa'] = taxa;
		jsonData['tipoReajuste'] = EstoqueReajusteCabecalhoDomain.setTipoReajuste(tipoReajuste);
		jsonData['justificativa'] = justificativa;
		
		var estoqueReajusteDetalheModelLocalList = []; 
		for (EstoqueReajusteDetalheModel object in estoqueReajusteDetalheModelList ?? []) { 
			estoqueReajusteDetalheModelLocalList.add(object.toJson); 
		}
		jsonData['estoqueReajusteDetalheModelList'] = estoqueReajusteDetalheModelLocalList;
		jsonData['viewPessoaColaboradorModel'] = viewPessoaColaboradorModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idColaborador = plutoRow.cells['idColaborador']?.value;
		dataReajuste = Util.stringToDate(plutoRow.cells['dataReajuste']?.value);
		taxa = plutoRow.cells['taxa']?.value?.toDouble();
		tipoReajuste = plutoRow.cells['tipoReajuste']?.value != '' ? plutoRow.cells['tipoReajuste']?.value : 'Aumentar';
		justificativa = plutoRow.cells['justificativa']?.value;
		estoqueReajusteDetalheModelList = [];
		viewPessoaColaboradorModel = ViewPessoaColaboradorModel();
		viewPessoaColaboradorModel?.nome = plutoRow.cells['viewPessoaColaboradorModel']?.value;
	}	

	EstoqueReajusteCabecalhoModel clone() {
		return EstoqueReajusteCabecalhoModel(
			id: id,
			idColaborador: idColaborador,
			dataReajuste: dataReajuste,
			taxa: taxa,
			tipoReajuste: tipoReajuste,
			justificativa: justificativa,
			estoqueReajusteDetalheModelList: estoqueReajusteDetalheModelListClone(estoqueReajusteDetalheModelList!),
		);			
	}

	estoqueReajusteDetalheModelListClone(List<EstoqueReajusteDetalheModel> estoqueReajusteDetalheModelList) { 
		List<EstoqueReajusteDetalheModel> resultList = [];
		for (var estoqueReajusteDetalheModel in estoqueReajusteDetalheModelList) {
			resultList.add(
				EstoqueReajusteDetalheModel(
					id: estoqueReajusteDetalheModel.id,
					idEstoqueReajusteCabecalho: estoqueReajusteDetalheModel.idEstoqueReajusteCabecalho,
					idProduto: estoqueReajusteDetalheModel.idProduto,
					valorOriginal: estoqueReajusteDetalheModel.valorOriginal,
					valorReajuste: estoqueReajusteDetalheModel.valorReajuste,
				)
			);
		}
		return resultList;
	}

	
}